pkg/proxy/backend/authenticator.go

```diff
@@ -24,74 +24,116 @@ import (
 	"github.com/pingcap/tidb/util/hack"
+	"go.uber.org/zap"
+)
+
+var (
+	ErrUnsupportedCapability = errors.New("unsupported capability")
 )
 
 // Other server capabilities are not supported.
-const supportedServerCapabilities = mysql.ClientLongPassword | mysql.ClientFoundRows | mysql.ClientLongFlag |
-	mysql.ClientConnectWithDB | mysql.ClientNoSchema | mysql.ClientODBC | mysql.ClientLocalFiles | mysql.ClientIgnoreSpace |
-	mysql.ClientProtocol41 | mysql.ClientInteractive | mysql.ClientSSL | mysql.ClientIgnoreSigpipe |
-	mysql.ClientTransactions | mysql.ClientReserved | mysql.ClientSecureConnection | mysql.ClientMultiStatements |
-	mysql.ClientMultiResults | mysql.ClientPluginAuth | mysql.ClientConnectAtts | mysql.ClientPluginAuthLenencClientData |
-	mysql.ClientDeprecateEOF
+const supportedServerCapabilities = pnet.ClientLongPassword | pnet.ClientFoundRows | pnet.ClientLongFlag |
+	pnet.ClientConnectWithDB | pnet.ClientLocalFiles | pnet.ClientProtocol41 | pnet.ClientInteractive | pnet.ClientSSL |
+	pnet.ClientTransactions | pnet.ClientSecureConnection | pnet.ClientMultiStatements |
+	pnet.ClientMultiResults | pnet.ClientPluginAuth | pnet.ClientConnectAttrs | pnet.ClientDeprecateEOF
 
 // Authenticator handshakes with the client and the backend.
 type Authenticator struct {
-	backendTLSConfig *tls.Config
-	dbname           string // default database name
-	serverAddr       string
-	user             string
-	attrs            []byte // no need to parse
-	capability       uint32 // client capability
-	collation        uint8
+	backendTLSConfig            *tls.Config
+	supportedServerCapabilities pnet.Capability
+	dbname                      string // default database name
+	serverAddr                  string
+	user                        string
+	attrs                       []byte // no need to parse
+	capability                  uint32 // client capability
+	collation                   uint8
 }
 
 func (auth *Authenticator) String() string {
 	return fmt.Sprintf("user:%s, dbname:%s, capability:%d, collation:%d",
 		auth.user, auth.dbname, auth.capability, auth.collation)
 }
 
-func (auth *Authenticator) handshakeFirstTime(clientIO, backendIO *pnet.PacketIO, frontendTLSConfig, backendTLSConfig *tls.Config) error {
+func (auth *Authenticator) handshakeFirstTime(logger *zap.Logger, clientIO, backendIO *pnet.PacketIO, salt []byte, frontendTLSConfig, backendTLSConfig *tls.Config, proxyProtocol bool) error {
+	clientIO.ResetSequence()
 	backendIO.ResetSequence()
-	// Read initial handshake packet from the backend.
-	serverPkt, serverCapability, err := auth.readInitialHandshake(backendIO)
-	if serverPkt != nil {
-		if writeErr := clientIO.WritePacket(serverPkt, true); writeErr != nil {
-			return writeErr
+
+	serverCapability := auth.supportedServerCapabilities
+	if frontendTLSConfig == nil {
+		serverCapability ^= pnet.ClientSSL
+	}
+	if err := clientIO.WriteInitialHandshake(serverCapability.Uint32(), salt, mysql.AuthCachingSha2Password); err != nil {
```

Comment:
Why does the proxy send a handshake packet first? I read the proxy protocol and I think the proxy can just forward the handshake from the server to the client.

Reply:
> some clients does not really respect the standard

Does it include AWS NLB?

> But some servers won't read the header lazily.

We only need to support TiDB. If TiDB supports reading lazily, then the proxy can connect to TiDB first and read the initial handshake.
TiProxy will never support other servers, because it depends on PD, `SHOW SESSION_STATES`, token based auth, etc.

> Of course, another reason is that per-namespace and other refactors relies on such behaviors.

We should respect how gateway routes connections. I was told it's an L4 proxy, so I don't think it generates an initial handshake by itself.

The reason why the proxy should not write the initial handshake packet by itself is that the handshake packet sent by the server may change, it includes salt, capability, and default auth plugin.

Reply:
> Does it include AWS NLB?

Yes.

> TiProxy will never support other servers, because it depends on PD, SHOW SESSION_STATES, token based auth, etc.

It could remove dependencies on PD, if dev tier GW and tiproxy merges as one. BTW, tiproxy could be used as a proxy in multi layer proxies. There could be lots of cases where tiproxy does not connect tidb directly.

> We should respect how gateway routes connections. I was told it's an L4 proxy.

Parsing and intercepting mysql proxy means it is already an L7 proxy. Can we remove `packetIO` from this package? Or can we do not migrate session by sending SQL that is not from clients? It is not even a L4 proxy in the respect of mysql protocol.

Also, handshaking clients first are the only way to do per-namespace routing, unless we route connections by public IP only.

> The reason why the proxy should not write the initial handshake packet by itself is that the handshake packet sent by the server may change, it includes salt, capability, and default auth plugin.

1. `capability` does matter, but not much. Most packets are intercepted by tiproxy, which means we could handle the case of different capabilities. Also, since we support same capabilities as tidb, there is no problem for now.
2. `auth` will be handled by `switch auth request`
3. `salt` or `plugin-auth-data` could be improved by caching. The next time handshake from the same user would know which backend to connect. In this case, we can forward initial handshake from backends directly. This has beyond the scope of this PR.

If you feel handshaking first is too dangerous, I could revert it in this PR: forward server `initialHandshake` in this PR. And do `handshake clients first + client/backend caching` in another PR.

Reply:
> salt or plugin-auth-data could be improved by caching.

I didn't mean the performance. The salt generated from the TiDB each time is different. TiProxy cannot generate a random salt if it doesn't authenticate the client by itself.

Reply:
> I didn't mean the performance. The salt generated from the TiDB each time is different. TiProxy cannot generate a random salt if it doesn't authenticate the client by itself.

I didn't mean performance too. Caching will enable us to choose the correct namespace for routing by pairing IP and other information. In other words, it gives a chance to just forward server initial handshake.

Reply:
Done, forwarding initial handshake first.